### PR TITLE
Code cleanup & fix the new jpc_floorlog2() assertion failure (#225)

### PR DESCRIPTION
**Pull request by MaxKellermann**

### CODE REVIEW

**Comment by osamu620** [comment]
In accordance to the tagtree data structure used in the JPEG 2000, my guess is the original author's intention is what @jubalh says.

**Comment by MaxKellermann** [comment]
> Only thing is a typo in the commit message `hjpc_tagtree: add unnecessary NULL check before jas_free()` where you probably wanted to write `remove` instead of `add`.

Fixed.

> And I wonder if the intention of the original author actually was

No. All callers have a `NULL` check. `free(NULL)` and `jas_free(NULL)` are legal, but `*_destroy(NULL)` is not. That's how all other `destroy` functions are implemented.

File: src/libjasper/base/jas_icc.c [hunk 1]
```diff
@@ -1620,8 +1622,9 @@ static int jas_iccputuint(jas_stream_t *out, unsigned n, jas_ulonglong val)
 
 static int jas_iccputsint(jas_stream_t *out, unsigned n, jas_longlong val)
 {
+	assert(val >= 0);
 	jas_ulonglong tmp;
-	tmp = (val < 0) ? (abort(), 0) : val;
+	tmp = (val < 0) ? 0 : val;
```
Comment (jubalh) [inline src/libjasper/base/jas_icc.c:1627]: val < 0 will now never be the case, will it? So I don't think we need `tmp`. Except maybe for different type `jas_longlong` vs `jas_ulonglong`.

**Comment by jubalh** [comment]
> All callers have a NULL check

They do have indeed. So that's good. Still wonder if this was his intention.

File: src/libjasper/base/jas_icc.c (hunk 1 above)
Comment (MaxKellermann) [reply src/libjasper/base/jas_icc.c:1627]: This commit tries to do least-intrusive refactoring - on old code which didn't make sense. *Everything* about this function is completely wrong! I have no idea what the author had in mind.
But my commit doesn't even try to fix anything about that function - it only refactors an `abort()` which also affects release builds to `assert()` (debug builds only), but keeps the negative cropping (which never worked).

Comment (jubalh) [reply src/libjasper/base/jas_icc.c:1627]: I see. GTM then.

**Comment by mdadams** [comment]
> > Only thing is a typo in the commit message `hjpc_tagtree: add unnecessary NULL check before jas_free()` where you probably wanted to write `remove` instead of `add`.
> 
> Fixed.
> 
> > And I wonder if the intention of the original author actually was
> 
> `free(NULL)` and `jas_free(NULL)` are legal, but `*_destroy(NULL)` is not. That's how all other `destroy` functions are implemented.

@MaxKellermann 
The comment about whether NULL is allowed in certain contexts sounds about right.  Although JasPer is written in C, I come from the C++ world.  The *_destroy functions loosely model destructors from the C++ world, and one cannot invoke a destructor with a nullptr.  The function jas_free is meant to have a similar interface as the C library function free.  So, jas_free should allow null pointers to be freed (which should have no effect).

**Comment by mdadams** [comment]
@MaxKellermann @jubalh 
Back in the day (i.e., over 20 years ago when JasPer was first developed), I was told that some implementations of free did not treat correctly handle free(0).  I doubt that such implementations are likely to exist today.  Anyways, this might provide some context as to why there are/were many places in the code that protected against free(0).  Using wrapper functions like jas_malloc/jas_free would have been a better solution, but these wrappers were only added much later when I wanted to be able to track memory usage in the library.  At least, this is my best recollection of why things are this way.  Does that help at all?

File: src/libjasper/base/jas_icc.c (hunk 1 above)
Comment (mdadams) [reply src/libjasper/base/jas_icc.c:1627]: > val < 0 will now never be the case, will it? So I don't think we need `tmp`. Except maybe for different type `jas_longlong` vs `jas_ulonglong`.

@MaxKellermann My best guess of what happened here is that there wasn't a function to handle reading signed long long values and I knew that negative values were not needed at the time.  So, I used the already existing function for unsigned data output and aborted if the value to be output was negative.  The intention was probably to fix this at a later point, and this never got fixed.  Is this theory consistent with what you see?  Again, it is my best guess.  And if this explanation is correct, yes, it is gross, but... Please keep in mind that JasPer was originally written by me alone as a side project while I was doing my Ph.D..  JasPer was not directly related to my Ph.D. and in many ways was very detrimental to my progress on my Ph.D..  So, I was under tremendous time pressure to get things working in JasPer, no matter how gross some implementation details might be. Some of the code that I wrote back then makes me cringe when I look at today, but the alternative (if I didn't cut some corners) would have been that I spent so much time on JasPer that I would never have finished by Ph.D..

**Pull request merged**